### ISSUE TITLE
Daily density product

### ISSUE THREAD

**Issue by mgalloy** [NCAR/ucomp-pipeline#172]

**Comment by jburkepile** [comment]
I took some notes at the meeting with Alin about deriving density:

[alin_work_density.docx](https://github.com/user-attachments/files/16881368/alin_work_density.docx)

**Comment by detoma** [comment]
The goal is to produce one or two density maps a day, for the average files.

The best strategy seems to use Alin's already created data cube in his python code, transform it into a save file, and run the interpolation to determine the density in IDL.

The only caveat is that th data cube will need to be recreated wen there is a relevant CHIANTI update

**Comment by detoma** [comment]
We plan to peak_intensity multiplied by the FWHM to derive the density. Shall we use the FWHM in nm?  Double check with Alin.

**Comment by mgalloy** [comment]
``` Python
>>> import bumpy as np
>>> f = "chianti_v10.1_pycelp_fe13_h99_d120_ratio.npz"
```

**Comment by mgalloy** [comment]
For a first example, try to compute density with these two files:

- `20220225.182056.ucomp.1074.l2.fts`
- `20220225.182341.ucomp.1079.l2.fts`

**Comment by mgalloy** [comment]
Here is a first draft of the density quicklook:

![20220225 182056-182341 ucomp 1074-1079 density](https://github.com/user-attachments/assets/80751db7-a713-4dca-ba2f-4c6d6fbb0f07)

I still need to:

- better annotation text
- refine `UCOMP_COMPUTE_DENSITY` to do bilinear interpolation

**Comment by mgalloy** [comment]
@detoma Does your ratio formula put the line width into units of nm? At least in the ratio, `c` cancels and that's why you have the 1074.4 and 1079.8 in the ratio?

**Comment by mgalloy** [comment]
The following interpolates in the same way that Alin does in his example, though also for height. Removed gamma correction, at least for now.

![20220225 182056-182341 ucomp 1074-1079 density](https://github.com/user-attachments/assets/9086e586-c03c-4d35-bf5e-2563f484af9f)

**Comment by mgalloy** [comment]
I think the following keywords, at least, are wrong from just taking the primary header of the 1074 file:

```
DATE-OBS= '2022-02-25T18:20:56.09' / [UT] date/time when obs started
DATE-END= '2022-02-25T18:23:21.96' / [UT] date/time when obs ended
MJD-OBS =      59635.764537038 / [days] modified Julian date
MJD-END =      59635.766215278 / [days] modified Julian date
FILTER  = '1074    '           / [nm] prefilter wavelength region identifier
OBJECT  = 'SUN     '           /
BUNIT   = '1.0E-06 B/Bsun'     / brightness with respect to solar disk
LEVEL   = 'L2      '           / level 2 calibrated
```

All the SGS and temperature values, most of the observing values are wrong too. Should I just remove them all?

**Comment by detoma** [comment]
Agreed. We can remove them. We can keep LEVEL as L2. I think it is fine to call densities level 2, unless we want to call them level 3.

**Comment by mgalloy** [comment]
By the way, [ucomp_convert_npz.py](analysis/ucomp_convert_npz.py) can convert an `.npz` file containing `h`, `den`, and `rat` produced by Alin's code to a netCDF file used by the pipeline. We could put more metadata into the netCDF file.

**Comment by mgalloy** [comment]
Current density GIF for:

- `20240409.191848.ucomp.1074.l2.fts`
- `20240409.191146.ucomp.1079.l2.fts`

is:

![20240409 191848-191146 ucomp 1074-1079 density](https://github.com/user-attachments/assets/32016978-2d98-4f26-a408-460180c08d2f)

**Comment by detoma** [comment]
There are faint bands that are noticeable in the density image especially in the N and S areas where w have low S/N. This is why there is no masking applied beside asking that intensities are > 0.

We are going to test different masking.  A first mask were we request that intensities are > 0.05 and a second man where we also add conditions on the FWHM and add minimum and maximum values to reject unphysical gaussian fits

The pixels rejected are the following:
259774 pixels for 0.0 intensity min
259213 pixels for 0.05 intensity min
257366 pixels for line width constraints

**Comment by mgalloy** [comment]
Masking a bit more with intensity minimums of 0.05 for both 1074 and 1079, and line width constraints from the wave region noise line width thresholds, gives the following:

![20240409 191848-191146 ucomp 1074-1079 density](https://github.com/user-attachments/assets/d96086de-eb08-4c64-a92b-14fffbdba762)

The number of good pixels for the various maskings:

- 259774 pixels for 0.0 intensity min
- 259213 pixels for 0.05 intensity min
- 257366 pixels for line width constraints

**Comment by detoma** [comment]
Mike is doing the interpolation slightly different from Alin. Alin chose the  largest heigh in the grid for each pixel. Mike uses both the lower and higher heights that bracket the height of the pixel height. 

We are going to test if the ratio is always in the range that coms out from CHIANTI or if sometimes, w are outside of the min/max ratio at a given height. In these cases, the pixel will be st to NaN.

**Comment by mgalloy** [comment]
For 20220111:

- `20220111.192841.ucomp.1074.l2.fts`
- `20220111.193101.ucomp.1079.l2.fts`

the density is:

![20220111 192841-193101 ucomp 1074-1079 density](https://github.com/user-attachments/assets/0b7075c1-c796-46c7-9efc-cb5b159c79b5)

**Comment by mgalloy** [comment]
Current primary header:

```
SIMPLE  =                    T / image conforms to FITS standard
BITPIX  =                  -32 / bits per data value
NAXIS   =                    2 / number of axes
NAXIS1  =                 1280 /
NAXIS2  =                 1024 /
EXTEND  =                    T / file may contain extensions
COMMENT --- Basic info ---
ORIGIN  = 'NCAR/HAO'           / Institution
INSTRUME= 'UCoMP'              / Upgraded Coronal Multichannel Polarimeter
TELESCOP= '20 cm One Shot'     / NSO One Shot telescope
LOCATION= 'MLSO'               / MLSO
OBSSWID = '1.0.6   '           / data collection software ID
DATE-OBS= '2022-01-11T19:28:41.72' / [UT] date/time when obs started
DATE-END= '2022-01-11T19:30:42.18' / [UT] date/time when obs ended
MJD-OBS =      59590.811585649 / [days] modified Julian date
MJD-END =      59590.812986112 / [days] modified Julian date
FILTER  = '1074    '           / [nm] prefilter wavelength region identifier
OBJECT  = 'SUN     '           /
BUNIT   = '1.0E-06 B/Bsun'     / brightness with respect to solar disk
LEVEL   = 'L2      '           / level 2 calibrated
COMMENT --- World Coordinate System (WCS) info ---
COMMENT Ephemeris calculations done by sun.pro
WCSNAME = 'helioprojective-cartesian' / World Coordinate System (WCS) name
CDELT1  =                2.834 / [arcsec/pixel] image X increment = platescale
CDELT2  =                2.834 / [arcsec/pixel] image Y increment = platescale
CRPIX1  =                640.5 / [pixel] occulter X center (index origin=1)
CTYPE1  = 'HPLN-TAN'           / helioprojective west angle: solar X
CRVAL1  =                 0.00 / [arcsec] occulter X sun center
CUNIT1  = 'arcsec  '           / unit of CRVAL1
CRPIX2  =                512.5 / [pixel] occulter Y center (index origin=1)
CTYPE2  = 'HPLT-TAN'           / helioprojective north angle: solar Y
CRVAL2  =                 0.00 / [arcsec] occulter Y sun center
CUNIT2  = 'arcsec  '           / unit of CRVAL2
DSUN_OBS=       147124599720.6 / [m] distance to the Sun from observer
HGLN_OBS=                0.000 / [deg] Stonyhurst heliographic longitude
HGLT_OBS=               -4.202 / [deg] Stonyhurst heliographic latitude
PC1_1   =                1.000 / coord transform matrix element (1, 1) WCS std.
PC1_2   =                0.000 / coord transform matrix element (1, 2) WCS std.
PC2_1   =                0.000 / coord transform matrix element (2, 1) WCS std.
PC2_2   =                1.000 / coord transform matrix element (2, 2) WCS std.
COMMENT --- Ephemeris info ---
COMMENT Ephemeris calculations done by sun.pro
SOLAR_P0=               -3.145 / [deg] solar P angle applied (image has N up)
SOLAR_B0=               -4.202 / [deg] solar B-Angle
SECANT_Z=             1.961522 / secant of the Zenith Distance
SID_TIME=              0.12562 / [day fraction] GMST sidereal time
CAR_ROT =                 2253 / Carrington Rotation Number
JUL_DATE=    2459591.311585649 / [days] Julian date
RSUN_OBS=               975.76 / [arcsec] solar radius using ref radius 959.63"
R_SUN   =               344.31 / [pixel] solar radius
COMMENT --- Level 1 processing info ---
DOI     = 'https://doi.org/10.26024/g8p7-wy42' / Digital Object Identifier
DATE_DP = '2023-11-30T06:24:08' / [UT] L1 processing date/time
DPSWID  = '1.0.1 [9f35649]'    / L1 processing software (2023-11-29) [production
LIN_CRCT=                    F / camera linearity corrected
DEMODV  =                    1 / demod coeffs version [2023-05-15T10:07:04Z]
CONTSUB =                    T / whether the continuum was subtracted
CAMERAS = 'both    '           / cameras used in processing
BOPAL   =                14.80 / [B/Bsun] opal radiance
COMMENT --- Level 2 processing info ---
DATE_DP2= '2023-11-30T09:03:14' / [UT] L2 processing date/time
DPSWID2 = '1.0.1 [9f35649]'    / L2 processing software (2023-11-29) [production
D_LAMBDA=                0.110 / [nm] wavelength spacing
COMMENT --- Quality metrics ---
VCROSSTK=             0.691477 / Stokes V crosstalk metric
MED_BKG =               13.753 / [ppm] median of line center background annulus
NUMSAT0O=                    0 / number of saturated pixels in onband RCAM
NUMSAT1O=                    0 / number of saturated pixels in onband TCAM
NUMSAT0C=                    0 / number of saturated pixels in bkg RCAM
NUMSAT1C=                    0 / number of saturated pixels in bkg TCAM
NUMNL0O =                   41 / number of non-linear pixels in onband RCAM
NUMNL1O =                   97 / number of non-linear pixels in onband TCAM
NUMNL0C =                   35 / number of non-linear pixels in bkg RCAM
NUMNL1C =                   88 / number of non-linear pixels in bkg TCAM
COMMENT --- Camera info ---
EXPTIME =               80.000 / [ms] Exposure time
FRAMERT =               10.691 / [Hz] Frequency of images
GAIN    = 'high'               / Camera gain setting
SAVEALL =                    F / [TF] Save all frames instead of summing
TCAMID  = 'Raptor OWL1280 10055' / Unique ID of the TCAM (Camera 1)
RCAMID  = 'Raptor OWL1280 10006' / Unique ID of the RCAM (Camera 0)
TCAMLUT = ''                   / Unique ID of the LUT for TCAM
RCAMLUT = ''                   / Unique ID of the LUT for RCAM
TCAMNUC = 'normal'             / Camera Correction for TCAM
RCAMNUC = 'normal'             / Camera Correction for RCAM
NFRAME  =                  160 / total number of image frames in file
REMFRAME=                    1 / number of bad frames removed
NUMWAVE =                    5 / number of wavelengths
NUMSUM  =                   16 / number of camera reads summed in an image frame
NREPEAT =                    2 / number of repeats of wavelength scans
NUMBEAM =                    2 / number of beams
COMMENT Total camera reads in this file = NFRAME * NUMSUM where
COMMENT   NFRAME = NUMWAVE * NREPEAT * NUMBEAM * 2(Cameras) * 4(Polarizations)
COMMENT --- Observing info ---
OBSERVER= 'Berkey'
OBS_ID  = '1074_05wave_2beam_16sums_2rep_BOTH' / Name of current observation
OBS_IDAU= ''                   / Author of the observing program
OBS_IDDA= ' Mon Sep 20 14:30:49 2021' / Edit date of the observing program
OBS_IDVE= '7f10e4c4e8f3f7b0072ee934b920f90f' / [Md5 hash] Observation ver
OBS_PLAN= 'all_wavelenght_coronal.cbk' / Name of Current program
OBS_PLAU= 'BERKEY'             / Author of the program
OBS_PLDA= ''                   / Edit date of the observing plan
OBS_PLVE= '6a81ac11c29ae24c9b7eb179e0177574' / [Md5 hash] Program ver
COMMENT --- Hardware settings ---
DARKID  = 'DARK SHUT 1'        / Dark Shutter ID
O1ID    = 'O1#2'               / Unique ID of objective lens
DIFFSRID= 'S1B'                / Unique ID of diffuser used (change to diffuser
OCCLTRID= '35'                 / Unique ID of the current occulter
FLCVNEG =               -7.000 / [V] Negative FeLC Drive voltage
FLCVPOS =                7.000 / [V] Positive FeLC Drive voltage
POLHOFF =            11021.000 / [counts] Offset in counts for the Cal Pololariz
REDHOFF =             8694.000 / [counts] Offset in counts for the Cal Retarder
T_COMPS =                    T / [TF] Lyot turning temperature compensation on
WAVOFF  =                1.870 / [nm] Tuning offset for Lyot filter
LCVRELX =                0.250 / [s] delay after LCVR tuning before data
FILTFWHM=                0.138 / [nm] Lyot FWHM
CONTOFF =              1.25625 / [nm] continuum offset
COMMENT Continuum can be "red", "blue", or "both": "both" gives equal weight to
COMMENT red and blue sides, "red" samples 90% red contimuum and 10% blue, "blue"
COMMENT samples 90% blue continuum and 10% red; the continuum position is offset
COMMENT from line center by the value of CONTOFF
CONTIN  = 'both    '           / [both/blue/red] location of continuum
OCCLTR-X=               69.950 / Occulter X position
OCCLTR-Y=               61.500 / Occulter Y position
O1FOCUS =               61.480 / O1 focus position
COMMENT --- Temperatures ---
COMMENT Temperatures used in the Lyot filter calibrations are low-pass
COMMENT filtered and reported in keywords that start with T_. The raw,
COMMENT unfiltered temperature values for recorded temperatures are recorded
COMMENT in keywords that begin with TU_.
T_RACK  =               22.804 / [C] Computer Rack Temp
T_LCVR1 =               34.560 / [C] Lyot LCVR1 Temp
T_LCVR2 =               34.509 / [C] Lyot LCVR2 Temp
T_LCVR3 =               34.585 / [C] Lyot LCVR3 Temp
T_LNB1  =               34.572 / [C] LiNb1 Temp
T_MOD   =               32.098 / [C] Modulator Temp
T_LNB2  =               34.477 / [C] LiNb2 Temp
T_LCVR4 =               34.447 / [C] Lyot LCVR4 Temp
T_LCVR5 =               34.189 / [C] Lyot LCVR5 Temp
T_BASE  =               30.968 / [C] Instrument Baseplate Temp
TU_RACK =               23.483 / [C] Computer Rack Temp Unfiltered
TU_LCVR1=               34.563 / [C] Lyot LCVR1 Temp Unfiltered
TU_LCVR2=               34.506 / [C] Lyot LCVR2 Temp Unfiltered
TU_LCVR3=               34.581 / [C] Lyot LCVR3 Temp Unfiltered
TU_LNB1 =               34.570 / [C] LiNb1 Temp Unfiltered
TU_MOD  =               32.086 / [C] Modulator Temp Unfiltered
TU_LNB2 =               34.479 / [C] LiNb2 Temp Unfiltered
TU_LCVR4=               34.458 / [C] Lyot LCVR4 Temp Unfiltered
TU_LCVR5=               34.234 / [C] Lyot LCVR5 Temp Unfiltered
TU_BASE =               31.088 / [C] Instrument Baseplate Temp Unfiltereed
TU_C0ARR=                4.866 / [C] Camera 0 Sensor array temp Unfiltered
TU_C0PCB=               28.000 / [C] Camera 0 PCB board temp Unfiltered
TU_C1ARR=                5.025 / [C] Camera 1 Sensor array temp Unfiltered
TU_C1PCB=               27.500 / [C] Camera 1 PCB board temp Unfiltered
COMMENT --- SGS info ---
SGSSCINT=              1.43230 / [arcsec] SGS scintillation seeing estimate
SGSDIMV =              8.73740 / [V] SGS Dim Mean
SGSDIMS =              0.00652 / [V] SGS Dim Std
SGSSUMV =              8.73885 / [V] SGS Sum Mean
SGSSUMS =              0.00388 / [V] SGS Sum Std
SGSRAV  =             -0.00005 / [V] SGS RA Mean
SGSRAS  =              0.00891 / [V] SGS RA Std
SGSDECV =             -0.00012 / [V] SGS DEC Mean
SGSDECS =              0.00851 / [V] SGS DEC Std
SGSLOOP =              1.00000 / SGS Loop Fraction
SGSRAZR =             47.00000 / [V] SGS RA zero point
SGSDECZR=             -6.00000 / [V] SGS DEC zero point
COMMENT --- Weather info ---
WNDSPD  =                8.000 / [mph] wind speed
WNDDIR  =               61.000 / [deg] wind direction
COMMENT --- Occulter centering info ---
COMMENT X/YOFFSET define position w.r.t. distortion corrected L0 images
XOFFSET0=                0.868 / [pixels] RCAM occulter x-offset from CRPIX1
YOFFSET0=               -8.407 / [pixels] RCAM occulter y-offset from CRPIX2
RADIUS0 =              355.518 / [pixels] RCAM occulter radius
FITCHI0 =             0.029087 / [pixels] chi-squared for RCAM center fit
XOFFSET1=                2.564 / [pixels] TCAM occulter x-offset from CRPIX1
YOFFSET1=               -8.819 / [pixels] TCAM occulter y-offset from CRPIX2
RADIUS1 =              356.049 / [pixels] TCAM occulter radius
FITCHI1 =             0.020361 / [pixels] chi-squared for TCAM center fit
POST_ANG=              183.500 / [deg] post angle CCW from north
RADIUS  =              355.784 / [pixels] occulter average radius
IMAGESCL=             2.818861 / [arcsec/pixels] image scale for this file
RCAMECC =               0.0213 / occulter eccentricity in RCAM
TCAMECC =               0.0166 / occulter eccentricity in TCAM
HISTORY
HISTORY Level 1 calibration and processing steps:
HISTORY   - quality check to determine if the file should be processed
HISTORY   - average level 0 data with same onband and wavelength
HISTORY   - apply dark correction
HISTORY   - apply gain correction
HISTORY   - camera corrections such as hot pixel correction
HISTORY   - demodulation
HISTORY   - distortion correction
HISTORY   - find the occulter position and radius
HISTORY   - subtract continuum
HISTORY   - remove hoizontal/vertical bands
HISTORY   - center images using occulter position and rotate to north up
HISTORY   - combine the cameras
HISTORY   - polarimetric correction
HISTORY   - correct for sky transmission
HISTORY   - update FITS keywords
```

**Comment by mgalloy** [comment]
Current density FITS file has the following extensions:

```
  1  Peak intensity [1074 nm]
  2  Line width (FWHM) [1074 nm]
  3  Peak intensity [1079 nm]
  4  Line width (FWHM) [1079 nm]
```

Current primary header of density file:

```
SIMPLE  =                    T / image conforms to FITS standard
BITPIX  =                  -32 / bits per data value
NAXIS   =                    2 / number of axes
NAXIS1  =                 1280 /
NAXIS2  =                 1024 /
EXTEND  =                    T / file may contain extensions
COMMENT --- Basic info ---
ORIGIN  = 'NCAR/HAO'           / Institution
INSTRUME= 'UCoMP'              / Upgraded Coronal Multichannel Polarimeter
TELESCOP= '20 cm One Shot'     / NSO One Shot telescope
LOCATION= 'MLSO'               / MLSO
OBSSWID = '1.0.6   '           / data collection software ID
DATE-OBS= '2022-04-07T18:10:26.68' / [UT] date/time when obs started
DATE-END= '2022-04-07T18:15:38.00' / [UT] date/time when obs ended
MJD-OBS =    2459677.257245371 / [days] modified Julian date
MJD-END =    2459677.260856482 / [days] modified Julian date
FILTER  = '1074/1079'          / [nm] prefilter wavelength region identifier
OBJECT  = 'SUN     '           /
BUNIT   = 'millionths of solar disk center intensity' /
LEVEL   = 'L3      '           / level 3 calibrated
PRODUCT = 'image   '           /
COMMENT --- World Coordinate System (WCS) info ---
COMMENT Ephemeris calculations done by sun.pro
WCSNAME = 'helioprojective-cartesian' / World Coordinate System (WCS) name
CDELT1  =                2.944 / [arcsec/pixel] image X increment = platescale
CDELT2  =                2.944 / [arcsec/pixel] image Y increment = platescale
CRPIX1  =                640.5 / [pixel] occulter X center (index origin=1)
CTYPE1  = 'HPLN-TAN'           / helioprojective west angle: solar X
CRVAL1  =                 0.00 / [arcsec] occulter X sun center
CUNIT1  = 'arcsec  '           / unit of CRVAL1
CRPIX2  =                512.5 / [pixel] occulter Y center (index origin=1)
CTYPE2  = 'HPLT-TAN'           / helioprojective north angle: solar Y
CRVAL2  =                 0.00 / [arcsec] occulter Y sun center
CUNIT2  = 'arcsec  '           / unit of CRVAL2
DSUN_OBS=       149756289951.6 / [m] distance to the Sun from observer
HGLN_OBS=                0.000 / [deg] Stonyhurst heliographic longitude
HGLT_OBS=               -6.155 / [deg] Stonyhurst heliographic latitude
PC1_1   =                1.000 / coord transform matrix element (1, 1) WCS std.
PC1_2   =                0.000 / coord transform matrix element (1, 2) WCS std.
PC2_1   =                0.000 / coord transform matrix element (2, 1) WCS std.
PC2_2   =                1.000 / coord transform matrix element (2, 2) WCS std.
COMMENT --- Ephemeris info ---
COMMENT Ephemeris calculations done by sun.pro
SOLAR_P0=              -26.264 / [deg] solar P angle applied (image has N up)
SOLAR_B0=               -6.155 / [deg] solar B-Angle
SECANT_Z=             2.066232 / secant of the Zenith Distance
SID_TIME=              0.30659 / [day fraction] GMST sidereal time
CAR_ROT =                 2256 / Carrington Rotation Number
JUL_DATE=    2459677.257245371 / [days] Julian date
RSUN_OBS=               958.61 / [arcsec] solar radius using ref radius 959.63"
R_SUN   =               325.62 / [pixel] solar radius
RSUN_REF=          696136000.0 / [m] solar radius at 1025.0 nm
COMMENT --- Density ---
CHIANTIV= '10.1    '           / Chianti version used to calculate lookup table
CHIANTIF= 'chianti_v10.1_pycelp_fe13_h99_d120_ratio.nc' / Chianti lookup table
1074FILE= '20220407.181026.ucomp.1074.l2.fts' / level 2 1074 filename
1079FILE= '20220407.181312.ucomp.1079.l2.fts' / level 2 1079 filename
```

**Comment by mgalloy** [comment]
The below compares Alin's vs Richard Morton's Chianti files. First, using Alin's file:

<!-- Uploading "20240409.180747-210322.ucomp.1074-1079.density.gif"... —>

Then Richard Morton's:

<!-- Uploading "20240409.180747-210322.ucomp.1074-1079.morton.density.gif"... —>

**Comment by mgalloy** [comment]
The following compares the density files using two different Chianti database files, different height and density grids, and inverted 1074/1079 ratios.

The `.nc` file created from Alin's file uses:

- 99 heights from 1.01 - 2.0 $R_{\odot}$
- 120 densities from $10^6$ to $10^{12}$
- ratio is 1074 / 1079

![Image](https://github.com/user-attachments/assets/b1b70407-6b22-4f36-be8e-6929e7d08020)

Richard Morton's `.sav` file uses:

- 36 heights from 1.05 - 1.4 $R_{\odot}$
- 81 densities from $10^5$ to $10^{13}$
- ratio is 1079 / 1074

![Image](https://github.com/user-attachments/assets/5fedf0cb-6f39-4fef-83b1-c4b89ac52a69)

**Comment by mgalloy** [comment]
Richard Morton's file gives larger results. The $log_{10}$ of the difference from $10^{4.08966}$  - $10^{8.03579}$ is:

**Comment by detoma** [comment]
This is what I get when I plot the ratio of the densities as a function of solar radius. The ups and downs at 1.10, 1.15, 1.20 ... solar radii could be an effect of the interpolation on a different grid, but there are other differences I do not quite understand. Richard Morton file gives densities that are about 25% larger. We need to identify the cause(s) of this. This is much larger than I expected.

<img width="868" alt="Image" src="https://github.com/user-attachments/assets/2f68142e-496f-4451-bad1-0e11ef2c4c68" />

What we could do is to create a file with the same densities and heights as the one of R. Morton and then check how they compare.

**Comment by detoma** [comment]
This issue was not updated in a while but a lot of work was done on it. Some is explained in issue #312.

I will try to give a summary of the key steps below.

**Comment by detoma** [comment]
When Mike tried to reproduce Alin's densities using pycelp, he obtained very different results (density were a lot lower). This triggered a lot of comparison to debug it, including comparing the densities from Mike with the ones from Richard's code, and the ones directly calculated from CHIANTI, which is how Zihao does it. 

We changed several parameters to see the effects on the densities, including: using different CHIANTI versions, different Tmax, turning on/off the limb-darkening correction, using different abundance files, changing the number of levels in the Fe atom, etc. Some of this is described in issue #312 and some is described below.

Initially, we could not reconcile Mike's computation with the other computations, no matter what we did. Clearly, the problem was on our side.  It turned out Mike was not using Alin's notebook to call pycelp but his own code and the heights were not set up correctly. This is from Mike's email on February 6 2025:

"I think I have found a major difference — the height input to the calculation is not measured from the sun center, so Alin uses 0.01 to 1.0 for his heights, but then writes 1.01 to 2.0 into the .pyz file. So I am trying a run subtracting 1 from the height just for the calculation."

Basically, pycelp expect the height to start at zero at the solar limb, while Mike was computing heights from the Sun's center. This caused our heights to be 1 Rsun larger and this is why the density were so much lower. After fixing the heights, things agreed well and we could reproduce Alin's results.

Please note that to avoid possible issues related to the creation of the density images themselves, the validation was done directly on the save files created using CHIANTI or pycelp, not on the final level 3 density products.

**Comment by mgalloy** [comment]
The density FITS file now looks like this:

```
level3$ ucomp cat 20240409.210752-210322.ucomp.1074-1079.normal.density.fts
  1  Density
  2  Peak intensity [1074 nm]
  3  Line width (FWHM) [1074 nm]
  4  Peak intensity [1079 nm]
  5  Line width (FWHM) [1079 nm]
```

**Comment by mgalloy** [comment]
If `ignore_linewidth` is set (the default now), the density FITS file looks like:

```
level3$ ucomp cat 20240409.210752-210322.ucomp.1074-1079.normal.density.fts
  1  Density
  2  Peak intensity [1074 nm]
  3  Peak intensity [1079 nm]
```

**Comment by detoma** [comment]
Thanks!

This looks fine. We need to think about what to put in the primary header
ext=0 vs the ext=1 header.

**Comment by detoma** [comment]
I tested several parameters used in the density computation and the effect they have, as described below. We decided to use pycelp for the computations instead of CHANTI or CLE because pycelp allows to use all the levels in FeXIII and includes the center-to-limb effect on the radiation field.

**Fe XIII LEVELS**
The current chianti version V11, includes 749levels for Fe XIII. Since we need to co put the save file with the ratios as a function of density and height only once, this is not an expensive computation. As discussed in Del Zanna & Dayananda 2025, the CLE implementation in FORWARD does not have sufficient sublevels to properly model the IR Fe XIII lines.

**LIMB darkening** 
CHIANTI does not include the limb darkening effect. This is likely non important for the EUV Fe Lines that are collision dominated. However, has a significant effect on the IR Fe lines that have a radiative component. Below is a plot showing results from pycelp, keeping all parameters the same but turning off the limb-darkening. The ratio of the density with and without limb-darkening is plotted and difference can be as large as 10%.

<img width="830" alt="Image" src="https://github.com/user-attachments/assets/a4deeeda-c549-4998-8243-1263f112717d" />

This was the rationale to use pyclep which implements the in the code radiation field as discussed in the book of Landi deli Innocenti ad Landolfi.

**CHIANTI version**
When we compared our derived densities with Richard Morton ones, we found large differences of 20% or more. We concluded they were caused by an old version of CHIANTI (version 8?) used by Richard. There was a significant upgrade in the Fe lines and CHIANTI v9 or larger should be used.


**T_peak**
CHIANTI and pycelp use a slightly different peak formation temperature,  i.e. 1778279 and 1676832. The difference is caused by a different interpolation scheme of the CHIANTI data. Pycelp is more accurate. The difference is negligible in both the line intensity and density derivation at low heights. Slightly "larger" differences at higher heights.




**Solar Spectrum**
Both CHIANTI and pycelp default to approximate the solar spectrum is to use a black body at 6000K. I compared some solar reference spectra (WHI reference spectrum, ATLAS 2004, SAO 2010, SSI) with a black body spectrum. The best fit in the IR at the UCOMP wavelength is given by a temperature of 5800K. So Mike modified the default in pycelp.

<img width="992" alt="Image" src="https://github.com/user-attachments/assets/ee85c717-4395-452e-aec9-b2fe3e078f06" />

The difference are significant only at low densities when the radiation filed matter most. For densities above 10^8.5 the effect is of the order of 1% or less. Note that UCoMP sees densities that typically range between 10^7.0 and 10^8.5,. rarely UCOMP sees density above 10^9 (maybe more common at solar max).

**Comment by detoma** [comment]
Improvements over Alin's interpolation to convert the line ratio to density:

We use a finer grid in densities and R_sun. The interpolation implemented by Mike uses the two heights that bracket the height value of a given pixel, while Alin's pick the closest height to the one of the pixel.
